### PR TITLE
Introduce `Rc::into_inner`, as a parallel to `Arc::into_inner`

### PR DESCRIPTION
**Pull request by joshtriplett**
Unlike `Arc`, `Rc` doesn't have the same race condition to avoid, but
maintaining an equivalent API still makes it easier to work with both
`Rc` and `Arc`.

### CODE REVIEW

File: library/alloc/src/rc/tests.rs [hunk 1]
```diff
@@ -151,6 +151,22 @@ fn try_unwrap() {
...
+    assert_eq!(Rc::into_inner(x), Some(3));
+
+    let x = Rc::new(4);
+    let y = Rc::clone(&x);
+    assert_eq!(Rc::into_inner(x), None);
+    assert_eq!(Rc::into_inner(y), Some(4));
+
+    let x = Rc::new(5);
+    let _w = Rc::downgrade(&x);
+    assert_eq!(Arc::into_inner(x), Some(5));
```
Comment (dtolnay) [inline library/alloc/src/rc/tests.rs:166]:
```suggestion
    assert_eq!(Rc::into_inner(x), Some(5));
```

**Review by dtolnay — CHANGES_REQUESTED** [review]

File: library/alloc/src/rc/tests.rs (hunk 1 above)
Comment (joshtriplett) [reply library/alloc/src/rc/tests.rs:166]: Argh, good catch, thank you.

**Review by dtolnay — APPROVED** [review]

File: library/alloc/src/rc/tests.rs [hunk 2]
```diff
@@ -151,6 +151,22 @@ fn try_unwrap() {
...
+    let y = Rc::clone(&x);
+    assert_eq!(Rc::into_inner(x), None);
+    assert_eq!(Rc::into_inner(y), Some(4));
+
+    let x = Rc::new(5);
+    let _w = Rc::downgrade(&x);
+    assert_eq!(Rc::into_inner(x), Some(5));
+}
+
+
```
Comment (steffahn) [inline library/alloc/src/rc/tests.rs:169]:
```suggestion

```
One less newline will make the formatting tidy check happy :-)

**Comment by joshtriplett** [comment]
@bors r=dtolnay

Thanks for the fix!

**Comment by steffahn** [comment]
I just noticed that the `Rc::into_inner` defined in this PR introduced its own feature gate “`rc_into_inner`” instead of using “`arc_into_inner`”, too. Was that intentional?

**Comment by joshtriplett** [comment]
On Mon, Mar 13, 2023 at 03:37:44AM -0700, Frank Steffahn wrote:
> I just noticed that the `Rc::into_inner` defined in this PR introduced its own feature gate “`rc_into_inner`” instead of using “`arc_into_inner`”, too. Was that intentional?

Yes, because I didn't want this PR to precommit to stabilizing
simultaneously.

**Comment by steffahn** [comment]
Okay. So does that mean the stabilization FCP in the tracking issue is only concerned with `Arc::into_inner` or not? (Because that’s not completely clear to me, reading the statement “I think `Arc::into_inner` makes sense to stabilize at this point. And `Rc::into_inner` is trivial.”.)

And in any case, the tracking issue should perhaps be updated to mention `#![feature(rc_into_inner)]`, too then?

**Comment by steffahn** [comment]
_(To readers via e-mail: There were some edits in my last response.)_

**Comment by joshtriplett** [comment]
> maybe the argument is simply that this makes it easier to have the option to stabilize just one of them without needing to re-label the function’s feature flag?

Exactly.

**Pull request merged**